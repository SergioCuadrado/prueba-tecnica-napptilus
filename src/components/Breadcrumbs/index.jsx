import { Link, useLocation } from 'react-router-dom'

import './styles.css'

export const Breadcrumbs = () => {
  const location = useLocation()

  return (
    <nav>
        <Link to="/" className={location.pathname === '/' ? 'breadcrumb-active' : 'breadcrumb-not-active'}>Home</Link>
        <span className="breadcrumb-arrow">&gt;</span>
        <Link to="/products" className={location.pathname.startsWith('/products') ? 'breadcrumb-active' : 'breadcrumb-not-active'}>Products</Link>
    </nav>
  )
}
