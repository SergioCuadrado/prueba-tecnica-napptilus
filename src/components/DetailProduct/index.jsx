import { useState } from 'react'
import { useAddProduct } from '../../hooks/useAddProduct'

export const DetailProduct = ({ product, descriptionProducts, optionsProduct }) => {
  const storageDefault = optionsProduct.storages.length === 1 ? optionsProduct.storages[0].name : ''
  const colorDefault = optionsProduct.colors.length === 1 ? optionsProduct.colors[0].code : ''

  const [actions, setActions] = useState({ colorCode: colorDefault, storageCode: storageDefault })
  const { getAddProductCart } = useAddProduct()
  if (!optionsProduct) return null

  const handleColorsChange = ({ target }) => {
    const colorCode = Number(target.name)
    setActions({
      ...actions,
      colorCode
    })
  }

  const handleStorageChange = ({ target }) => {
    const storageCode = target.value
    setActions({
      ...actions,
      storageCode
    })
  }

  const handleSubmit = (e) => {
    e.preventDefault()
    const productAdded = {
      id: product.description.id,
      ...actions
    }
    getAddProductCart(productAdded)
  }

  return (
    <div className="container">
      <div className="imgBx" data-content={product.description.brand}>
          <img src={product.description.image} alt={product.description.model} />
      </div>
      <div className="details">
          <div className="content">
            <h2>{product.description.model}</h2>
            <table>
                <tbody>
                  {descriptionProducts
                    ? descriptionProducts.map((description) => (
                    <tr key={description.name}>
                      <td>
                        <span>
                          {description.name}
                        </span>
                      </td>
                      <td>
                        <span>
                          {description.value.length > 0 ? description.value : 'No especificado'}
                        </span>
                      </td>
                    </tr>
                    ))
                    : null}
                </tbody>
            </table>
            <form className='buy-product' onSubmit={handleSubmit}>
                <label className='capacity' htmlFor="capacity">Capacidad</label>
                <select name="capacity" id="capacity" onChange={handleStorageChange}>
                    {optionsProduct.storages.map((capacity) => (
                        <option key={capacity.code} value={capacity.name}>{capacity.name}</option>
                    ))}
                </select>
                <h4 className='colors-title'>Colores disponibles:</h4>
                <div className="product-colors" >
                {optionsProduct.colors.map((color) => (
                    <div key={color.code}>
                        <label className={`${color.name} ${actions.colorCode === color.code ? 'active' : null}`} style={{ backgroundColor: color?.name.toLowerCase() }} htmlFor={color.code} />
                        <input type="checkbox" name={color.code} id={color.code} onChange={handleColorsChange} checked={actions.color === color.code} hidden />
                    </div>
                ))}
                </div>
                <h3>$ {product.description.price}</h3>
                <button disabled={actions.colorCode === '' || actions.storageCode === ''}>Comprar</button>
            </form>
          </div>
      </div>
    </div>
  )
}
