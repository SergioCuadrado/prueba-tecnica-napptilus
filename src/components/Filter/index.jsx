import React, { useId } from 'react'
import { useFiltersProducts } from '@/hooks/useFiltersProducts'

import './styles.css'

export const Filter = () => {
  const searchFilter = useId()
  const { filter, setProductsFilter } = useFiltersProducts()

  const handleSubmit = (event) => {
    event.preventDefault()
  }

  const handleChange = ({ target }) => {
    const search = target.value
    setProductsFilter((prevState) => ({
      ...prevState,
      search
    }))
  }

  return (
    <>
        <form className="form" onSubmit={handleSubmit}>
          <label htmlFor={searchFilter}>BÃºsqueda por marca / modelo:</label>
          <input
              name={searchFilter}
              id={searchFilter}
              type="text"
              onChange={handleChange}
              value={filter.search}
              placeholder="Liquid, Iconia..."
              autoFocus
          />
        </form>
    </>
  )
}
