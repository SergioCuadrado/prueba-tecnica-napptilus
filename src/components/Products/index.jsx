import { useProducts } from '@/hooks/useProducts'
import { useEffect } from 'react'
import './styles.css'

export const Products = () => {
  const { products, loading, error, getProducts } = useProducts()

  useEffect(() => {
    getProducts()
  }, [])

  console.log({ products, loading, error })
  return (
    <div >
      <ul className='products'>
          {products.map((product) => (
              <li className='product' key={product.id}>
                  <img src={product.image} alt={product.model} />
                  <h4>{product.brand} - {product.model}
                  {product?.price.length > 0 ? <span>${product.price}</span> : null}
                  </h4>
              </li>
          ))}
      </ul>
    </div>
  )
}
