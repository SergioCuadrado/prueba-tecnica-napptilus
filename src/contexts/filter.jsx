import { createContext, useState } from 'react'

export const FilterContext = createContext()

export const ProductsFilterProvider = ({ children }) => {
  const [filter, setProductsFilter] = useState({
    search: ''
  })

  return (
    <FilterContext.Provider value={{ filter, setProductsFilter }}>
      {children}
    </FilterContext.Provider>
  )
}
