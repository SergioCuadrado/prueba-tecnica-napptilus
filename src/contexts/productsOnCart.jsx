import { createContext, useState } from 'react'

export const ProductsOnCartContext = createContext()

export const ProductsOnCartProvider = ({ children }) => {
  const [productsOnCart, setProductsOnCart] = useState(0)

  return (
    <ProductsOnCartContext.Provider value={{ productsOnCart, setProductsOnCart }}>
      {children}
    </ProductsOnCartContext.Provider>
  )
}
