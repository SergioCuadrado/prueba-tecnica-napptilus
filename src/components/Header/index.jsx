import { Link } from 'react-router-dom'
import { Breadcrumbs } from '../Breadcrumbs'
import { CartIcon } from '../Icons'
import './styles.css'

export const Header = () => {
  return (
    <header>
      <div>
        <Link to='/' className='title-app'>
          <h1>Napptilus-Shop</h1>
        </Link>
        <Breadcrumbs />
      </div>
      <label className='cart-button' htmlFor="">
        <CartIcon />
        <span>5</span>
      </label>
    </header>
  )
}
