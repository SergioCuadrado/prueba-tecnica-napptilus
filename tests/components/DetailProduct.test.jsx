import { render, screen, fireEvent } from '@testing-library/react'
import { DetailProduct } from '@/components/DetailProduct'
import { ProductsOnCartContext } from '@/contexts/productsOnCart'

describe('Pruebas en el componente <DetailProduct />', () => {
  const product = {
    description: {
      battery: 'Non-removable Li-Ion 3400 mAh battery (12.92 Wh)',
      brand: 'Acer',
      cpu: 'Quad-core 1.3 GHz Cortex-A53',
      dimensions: '191.7 x 101 x 9.4 mm (7.55 x 3.98 x 0.37 in)',
      id: 'ZmGrkLRPXOTpxsU4jjAcv',
      image: 'https://itx-frontend-test.onrender.com/images/ZmGrkLRPXOTpxsU4jjAcv.jpg',
      model: 'Iconia Talk S',
      os: 'Android 6.0 (Marshmallow)',
      price: '170',
      primaryCamera: ['13 MP', 'autofocus'],
      ram: '2 GB RAM',
      screenResolution: '7.0 inches (~69.8% screen-to-body ratio)',
      secondaryCamera: ['2 MP', '720p'],
      weight: '260'
    },
    options: {
      colors: [
        {
          code: 1,
          name: 'Black'
        },
        {
          code: 2,
          name: 'White'
        }
      ],
      storages: [
        {
          code: 1,
          name: '26 GB'
        },
        {
          code: 2,
          name: '32 GB'
        }
      ]
    }
  }

  const descriptionProducts = [
    { name: 'Battery', value: 'Non-removable Li-Ion 3400 mAh battery (12.92 Wh)' },
    { name: 'Modelo', value: 'Iconia Talk S' },
    { name: 'Precio', value: '170' },
    { name: 'CPU', value: 'Quad-core 1.3 GHz Cortex-A53' },
    { name: 'RAM', value: '2 GB RAM' },
    { name: 'Sistema Operativo', value: 'Android 6.0 (Marshmallow)' },
    { name: 'Resolución de pantalla', value: '7.0 inches (~69.8% screen' },
    { name: 'Batería', value: 'Non-removable Li-Ion 3400' },
    { name: 'Cámara principal', value: ['13 MP', 'autofocus'] },
    { name: 'Cámara secundaria', value: ['2 MP', '720p'] },
    { name: 'Dimensiones', value: '191.7 x 101' },
    { name: 'Peso', value: '260' }
  ]

  const optionsProduct = {
    colors: [
      {
        code: 1,
        name: 'Black'
      },
      {
        code: 2,
        name: 'White'
      }
    ],
    storages: [
      {
        code: 1,
        name: '16 GB'
      },
      {
        code: 2,
        name: '32 GB'
      }
    ]
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('Debe de mostrar el componente correctamente', () => {
    const productsOnCart = 0
    const setProductsOnCartMock = jest.fn()

    const { container } = render(
        <ProductsOnCartContext.Provider value={{ productsOnCart, setProductsOnCart: setProductsOnCartMock }} >
            <DetailProduct product={product} descriptionProducts={descriptionProducts} optionsProduct={optionsProduct} />
        </ProductsOnCartContext.Provider>)
    expect(container).toMatchSnapshot()
  })

  test('Si algun campo del formulario no esta seleccionado que el boton este desactivado', () => {
    const productsOnCart = 0
    const setProductsOnCartMock = jest.fn()

    render(
        <ProductsOnCartContext.Provider value={{ productsOnCart, setProductsOnCart: setProductsOnCartMock }} >
            <DetailProduct product={product} descriptionProducts={descriptionProducts} optionsProduct={optionsProduct} />
        </ProductsOnCartContext.Provider>
    )

    const button = screen.getByRole('button', { name: /añadir a la cesta/i })
    expect(button).toBeDisabled()
  })

  test('Que funcione correctamente el formulario', async () => {
    const productsOnCart = 0
    const setProductsOnCartMock = jest.fn()

    const optionsProduct = {
      colors: [
        {
          code: 1,
          name: 'Black'
        },
        {
          code: 2,
          name: 'White'
        }
      ],
      storages: [
        {
          code: 1,
          name: '16 GB'
        }
      ]
    }

    render(
        <ProductsOnCartContext.Provider value={{ productsOnCart, setProductsOnCart: setProductsOnCartMock }} >
            <DetailProduct product={product} descriptionProducts={descriptionProducts} optionsProduct={optionsProduct} />
        </ProductsOnCartContext.Provider>
    )
    // comprobar que el boton del formulario este desactivado
    const button = screen.getByRole('button', { name: /Añadir a la cesta/i })
    expect(button).toBeDisabled()

    const color = screen.getByLabelText('color-1')
    fireEvent.click(color)

    // comprobar que el boton del formulario este activado
    expect(button).not.toBeDisabled()
    fireEvent.click(button)
  })

  test('Que si solo tienen un valor por cada campo del formulario el boton este activo desde el inicio', () => {
    const productsOnCart = 0
    const setProductsOnCartMock = jest.fn()

    const optionsProduct = {
      colors: [
        {
          code: 1,
          name: 'Black'
        }
      ],
      storages: [
        {
          code: 1,
          name: '16 GB'
        }
      ]
    }

    render(
      <ProductsOnCartContext.Provider value={{ productsOnCart, setProductsOnCart: setProductsOnCartMock }} >
          <DetailProduct product={product} descriptionProducts={descriptionProducts} optionsProduct={optionsProduct} />
      </ProductsOnCartContext.Provider>
    )

    const button = screen.getByRole('button', { name: /Añadir a la cesta/i })
    expect(button).not.toBeDisabled()
  })
})
