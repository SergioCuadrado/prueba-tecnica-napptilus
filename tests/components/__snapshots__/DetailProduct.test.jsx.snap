// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <DetailProduct /> Debe de mostrar el componente correctamente 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="container"
  >
    <div
      class="imgBx"
      data-content="Acer"
    >
      <img
        alt="Iconia Talk S"
        src="https://itx-frontend-test.onrender.com/images/ZmGrkLRPXOTpxsU4jjAcv.jpg"
      />
    </div>
    <div
      class="details"
    >
      <div
        class="content"
      >
        <h2>
          Iconia Talk S
        </h2>
        <table>
          <tbody>
            <tr>
              <td>
                <span>
                  Battery
                </span>
              </td>
              <td>
                <span>
                  Non-removable Li-Ion 3400 mAh battery (12.92 Wh)
                </span>
              </td>
            </tr>
            <tr>
              <td>
                <span>
                  Modelo
                </span>
              </td>
              <td>
                <span>
                  Iconia Talk S
                </span>
              </td>
            </tr>
            <tr>
              <td>
                <span>
                  Precio
                </span>
              </td>
              <td>
                <span>
                  170
                </span>
              </td>
            </tr>
            <tr>
              <td>
                <span>
                  CPU
                </span>
              </td>
              <td>
                <span>
                  Quad-core 1.3 GHz Cortex-A53
                </span>
              </td>
            </tr>
            <tr>
              <td>
                <span>
                  RAM
                </span>
              </td>
              <td>
                <span>
                  2 GB RAM
                </span>
              </td>
            </tr>
            <tr>
              <td>
                <span>
                  Sistema Operativo
                </span>
              </td>
              <td>
                <span>
                  Android 6.0 (Marshmallow)
                </span>
              </td>
            </tr>
            <tr>
              <td>
                <span>
                  Resolución de pantalla
                </span>
              </td>
              <td>
                <span>
                  7.0 inches (~69.8% screen
                </span>
              </td>
            </tr>
            <tr>
              <td>
                <span>
                  Batería
                </span>
              </td>
              <td>
                <span>
                  Non-removable Li-Ion 3400
                </span>
              </td>
            </tr>
            <tr>
              <td>
                <span>
                  Cámara principal
                </span>
              </td>
              <td>
                <span>
                  13 MP
                  autofocus
                </span>
              </td>
            </tr>
            <tr>
              <td>
                <span>
                  Cámara secundaria
                </span>
              </td>
              <td>
                <span>
                  2 MP
                  720p
                </span>
              </td>
            </tr>
            <tr>
              <td>
                <span>
                  Dimensiones
                </span>
              </td>
              <td>
                <span>
                  191.7 x 101
                </span>
              </td>
            </tr>
            <tr>
              <td>
                <span>
                  Peso
                </span>
              </td>
              <td>
                <span>
                  260
                </span>
              </td>
            </tr>
          </tbody>
        </table>
        <form
          aria-label="form"
          class="buy-product"
          data-testid="form-buy-product"
        >
          <h4
            class="capacity"
          >
            Capacidad
          </h4>
          <div
            class="basic-single css-b62m3t-container"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-2-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
            />
            <div
              class=" css-v68sna-control"
            >
              <div
                class=" css-1fdsijx-ValueContainer"
              >
                <div
                  class=" css-1jqq78o-placeholder"
                  id="react-select-2-placeholder"
                >
                  Selecciona la capacidad de almacenamiento que desea
                </div>
                <div
                  class=" css-qbdosj-Input"
                  data-value=""
                >
                  <input
                    aria-autocomplete="list"
                    aria-describedby="react-select-2-placeholder"
                    aria-expanded="false"
                    aria-haspopup="true"
                    aria-label="Capacidad"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    class=""
                    id="react-select-2-input"
                    role="combobox"
                    spellcheck="false"
                    style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class=" css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class=" css-1u9des2-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class=" css-1xc3v61-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
          <h4
            class="colors-title"
          >
            Colores disponibles:
          </h4>
          <div
            class="product-colors"
          >
            <div>
              <label
                class="Black null"
                for="1"
                style="background-color: black;"
              />
              <input
                aria-label="color-1"
                hidden=""
                id="1"
                name="1"
                type="checkbox"
              />
            </div>
            <div>
              <label
                class="White null"
                for="2"
                style="background-color: white;"
              />
              <input
                aria-label="color-2"
                hidden=""
                id="2"
                name="2"
                type="checkbox"
              />
            </div>
          </div>
          <h3>
            $ 
            170
          </h3>
          <button
            disabled=""
          >
            Añadir a la cesta
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
